editoast/src/views/v2/path/projection.rs

```diff
@@ -0,0 +1,471 @@
+            length: pos,
+        }
+    }
+
+    /// Get the position in the path given a location (track section and offset).
+    pub fn get_position(&self, location: &TrackLocation) -> Option<u64> {
+        let base_position = *self.map_position.get(&location.track_section)?;
+        let track_range = self
+            .get_track_range(&location.track_section)
+            .expect("Track range should be found since we found the base_position");
+        // Check if the offset is in the range.
+        if location.offset < track_range.begin || location.offset > track_range.end {
+            return None;
+        }
+        if track_range.direction == Direction::StartToStop {
+            Some(base_position + location.offset - track_range.begin)
+        } else {
+            Some(base_position + track_range.end - location.offset)
+        }
+    }
+
+    /// Get the location (track section and offset) given a position in a path.
+    /// Is function uses a binary search to find the track section and offset.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the position is out of bounds.
+    pub fn get_location(&self, position: u64) -> TrackLocationFromPath {
+        assert!(position <= self.length, "Position out of bounds");
+
+        // Binary search to retrieve the corresponding track section range.
+        let mut left = 0;
+        let mut right = self.path.len() - 1;
+        while left != right {
+            let mid = (left + right + 1) / 2;
+            let mid_track_range = &self.path[mid];
+            let mid_position = self.map_position[&mid_track_range.track_section];
+
+            if mid_position > position {
+                right = mid - 1;
+            } else {
+                left = mid;
+            }
+        }
+
+        // Retrieve the first location and check if it's on another one
+        let index = left;
+        let found_track_range = &self.path[index];
+        let found_position = self.map_position[&found_track_range.track_section];
+        let mut has_next = false;
+        let mut has_prev = false;
+        let first_track_loc = if found_track_range.direction == Direction::StartToStop {
+            let offset = found_track_range.begin + position - found_position;
+            if offset == found_track_range.begin && index > 0 {
+                has_prev = true;
+            } else if offset == found_track_range.end && index < self.path.len() - 1 {
+                has_next = true;
+            }
+            TrackLocation::new(&found_track_range.track_section, offset)
+        } else {
+            let offset = found_track_range.end - (position - found_position);
+            if offset == found_track_range.end && index > 0 {
+                has_prev = true;
+            } else if offset == found_track_range.begin && index < self.path.len() - 1 {
+                has_next = true;
+            }
+            TrackLocation::new(&found_track_range.track_section, offset)
+        };
```

Comment:
From what I'm reading there's some chunks that appear to have some complex logic that could be made more readable with a good set of primitives. Here for example, if we had:

```rust
impl TrackRange {
    /// Computes a TrackRangeOffset location on this track range following its direction
    fn offset(&'a self, offset: i64) -> TrackRangeOffset<'a>
    fn length(&self) -> u64;
}

impl TrackRangeOffset<'track_range> {
   fn as_track_offset(&self) -> TrackOffset;
}
```

I believe this block could then be rewritten as:

```suggestion
        let first_track_loc = {
            let offset_loc = found_track_range.offset(position - found_position);
            if offset_loc.offset == 0 && index > 0 {
                has_prev = true;
            } else if offset_loc.offset == found_track_range.length() && index < self.path.len() - 1 {
                has_next = true;
            }
            offset_loc.as_track_offset()
        };
```